theano/gpuarray/opt.py

```diff
@@ -376,9 +376,20 @@ def apply(self, fgraph):
             if new_o.type != o.type:
                 assert isinstance(o.type, tensor.TensorType)
                 assert isinstance(new_o.type, GpuArrayType)
-                new_o = host_from_gpu(new_o)
+                assert new_o.owner
+                # This condition is needed in the case one input is an
+                # output of the graph. Without this, it would
+                # introduce cycle as we don't replace correctly that
+                # case. It would also add extra transfer to/from the
+                # gpu.
+                if (isinstance(new_o.owner.op, GpuFromHost) and
```

Comment:
It was not a test. That was a theano function that i tried compiling with the input and outputs of the graphs being same. I have sent the code on hangouts and i tested that again. It is fine now 👍